src/ngx_http_lua_resolve.c

```diff
@@ -0,0 +1,483 @@
+        ngx_http_lua_finalize_request(r, NGX_DONE);
+        return ngx_http_lua_run_posted_threads(c, vm, r, ctx, nreqs);
+    }
+
+    if (ctx->entered_content_phase) {
+        ngx_http_lua_finalize_request(r, rc);
+        return NGX_DONE;
+    }
+
+    return rc;
+}
+
+
+static void
+ngx_http_lua_resolve_cleanup(void *data)
+{
+    ngx_resolver_ctx_t            *rctx;
+    ngx_http_lua_co_ctx_t         *coctx = data;
+    ngx_http_upstream_resolved_t  *resolved;
+
+    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
+                   "lua ngx.resolve abort resolver");
+
+    resolved = coctx->data;
+    if (resolved == NULL) {
+        return;
+    }
+
+    rctx = resolved->ctx;
+    if (rctx == NULL) {
+        return;
+    }
+
+    /* just to be safer */
+    rctx->handler = ngx_http_lua_resolve_empty_handler;
+
+    ngx_resolve_name_done(rctx);
+}
+
+
+static void
+ngx_http_lua_resolve_empty_handler(ngx_resolver_ctx_t *ctx)
+{
+    /* do nothing */
+}
+
+
+static int
+ngx_http_lua_resolve_select_retval(ngx_http_lua_resolve_ctx_t *u, lua_State *L)
+{
+    ngx_uint_t                     i;
+    ngx_resolver_addr_t           *addr;
+    ngx_http_upstream_resolved_t  *ur = u->resolved;
+
+    for (i = 0; i < ur->naddrs; i++) {
+        addr = &ur->addrs[i];
+        switch (addr->sockaddr->sa_family) {
+#if (NGX_HAVE_INET6)
+            case AF_INET6:
+                if (u->ipv6) {
+                    return ngx_http_lua_resolve_get_retval(addr->sockaddr,
+                                                           addr->socklen, L);
+                }
+                break;
+#endif
+            default: /* AF_INET */
+                if (u->ipv4) {
+                    return ngx_http_lua_resolve_get_retval(addr->sockaddr,
+                                                           addr->socklen, L);
+                }
+        }
+    }
+
+    lua_pushnil(L);
+    lua_pushliteral(L, "address not found");
+
+    return 2;
+}
+
+
+static int
+ngx_http_lua_resolve_get_retval(struct sockaddr *sockaddr, socklen_t socklen,
+    lua_State *L)
+{
+    size_t  len;
+    u_char  text[NGX_SOCKADDR_STRLEN];
+
+    len = ngx_sock_ntop(sockaddr, socklen, text, NGX_SOCKADDR_STRLEN, 0);
+    lua_pushlstring(L, (const char *) text, len);
+
+    return 1;
+}
+
+
+static int
+ngx_http_lua_resolve_set_query_option(const char *name, unsigned *option, lua_State *L)
+{
+    lua_getfield(L, 2, name);
+
+    switch (lua_type(L, -1)) {
+        case LUA_TNIL:
+            /* do nothing */
+            break;
+
+        case LUA_TBOOLEAN:
+            *option = lua_toboolean(L, -1);
+            break;
+
+        default:
+            return luaL_error(L, "bad \"%s\" option value type: %s",
+                       name, luaL_typename(L, -1));
+    }
+
+    lua_pop(L, 1);
+
+    return NGX_OK;
+}
+
+
+void
+ngx_http_lua_inject_resolve_api(lua_State *L)
+{
+    lua_pushcfunction(L, ngx_http_lua_ngx_resolve);
+    lua_setfield(L, -2, "resolve");
```

Comment:
We no longer add new API functions to the global `ngx` table. The current way is to add pure Lua modules to lua-resty-core for any new API functions. See `ngx.pipe` module in that lua-resty-core library for an example.

Reply:
Got it, I have completely rewritten my initial implementation.